[Meowloid] iP

Meowloid opened the pull request:
# Majimabot
>"He's alright." - Ono Michio

## What is Majimabot?
This portable instance of the Mad Dog of Shimano can't fight for you, but what he lacks in combat skills, he makes up for in clerical tasks!

What can he do, you ask?

1. Keep track of various tasks
    * 'Todo', tasks which are either done or not
    * 'Deadline' tasks where you can specify a date and time to accomplish it by
    * 'Event' tasks which take place during two points in time
 2. Mark tasks as completed or not
 3. Find specific tasks you have listed
 
 ... and more to come! <sub>(If I feel like it.)</sub>

## How _do_ I use Majimabot? 🤔 

Majima tracks three types of tasks, as stated above.
Typing 'todo' in the console, followed by a short description, stores a task with the corresponding description.
Entering 'deadline', followed by a description, seperated by a '/by' and a valid date-time format of "dd-MM-YYYY HHmm" creates a task with a deadline that Majima will keep track of.
An event has a description, a /from and /to time which does not need to follow a date-time format.

To know what tasks Majima is currently keeping track of, simply tell Majima
`list`
and he'll show you what tasks he's currently remembering.

Furthermore, tasks can be marked or unmarked to keep track of their completion. In a list like such,
```
1. [T][ ] Beat up thugs
2. [T][ ] Random sidequest
3. [T][ ] Throw the trash
```
executing `mark 3` would mark the task like so:
```
1. [T][ ] Beat up thugs
2. [T][ ] Random sidequest
3. [T][X] Throw the trash
```
## So, what are you waiting for!
Majimabot remembers tasks in between runs, so download the .jar [here](https://github.com/Meowloid/ip/releases/tag/A-Jar), put him in a directory of your choice! And it's all for **FREE**!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Meowloid committed:
Level-0 implemented here, don't forget to tag

Meowloid committed:
personality change and level 1

Meowloid committed:
level 2

Meowloid committed:
Level-3, as well as some changes to make it look more like the testcases on the website

Meowloid committed:
New classes added to reflect different types of tasks and to enable more types of classes to be added, as per requirement. Level 4

Meowloid committed:
changes to enable A-TextUiTesting for this current level of Majima bot.

Meowloid committed:
MajimaDEPRECATED is old code before refactoring, as implementing exceptions with everything in main is a nightmare.
Level-5 implemented.

Meowloid committed:
What a day. Time to eat dinner.
Level-6

Meowloid committed:
enumerations added

Meowloid committed:
stop tracking old ver

Meowloid committed:
feature1 testing

Meowloid committed:
feature2 added

Meowloid committed:
Merge branch 'master' into feature1

Meowloid committed:
step 7

Meowloid committed:
step 8

Meowloid committed:
Merge branch 'feature1'

Meowloid committed:
countries.txt: add Singapore

Meowloid committed:
Merge branch 'add-countries'

Meowloid committed:
Add first boot file creation

Meowloid committed:
Add changes to Task and subclasses to facilitate saving to text file

Meowloid committed:
Add save functionality

Meowloid committed:
Change logic so that deleting rebuilds the .txt file so the delete call works as intended

Meowloid committed:
Merge branch 'branch-Level-7'

Meowloid committed:
Change Deadline task type to use LocalDateTime type instead of String for deadline.

Meowloid committed:
Update addDeadline method to handle invalid date/time format.
Update loadLine to be able to handle corrupted data with regards to invalid date/time format.

Meowloid committed:
Revert "Update addDeadline method to handle invalid date/time format."

This reverts commit ba5640ad352fe37c103d64d1df50f5cfdfd20743.

Meowloid committed:
Reapply "Update addDeadline method to handle invalid date/time format."

This reverts commit a209fc0e79cc279f7ee4f19a73bdb9879a7f7481.

Meowloid committed:
Merge branch 'branch-Level-8'

Meowloid committed:
Remove unncessary files unrelated to iP which were made by the weekly tasks and refactor Majima.java to adhere closely to OOP.

Meowloid committed:
Merge branch 'branch-A-MoreOOP'

Meowloid committed:
Put all class files together so Majima.java runs

Meowloid committed:
Put all classes into relevant folders and packages

Meowloid committed:
Merge branch 'add-gradle-support'

Meowloid committed:
Fix loading deadlines from .txt file and add gradle support

Meowloid committed:
Merge branch 'add-gradle-support'

Meowloid committed:
Add DeadlineTest and EventTest classes

Meowloid committed:
Merge branch 'branch-A-JUnit'

Meowloid committed:
Change build.gradle to support shadow plugin

Meowloid committed:
Merge branch 'branch-A-Jar'

Meowloid committed:
Add JavaDoc comments to all classes and public methods

Meowloid committed:
Add the FindCommand() method and 'find' functionality

Meowloid committed:
Merge branch 'branch-A-JavaDoc'

Meowloid committed:
Add checkstyle support in gradle

Meowloid committed:
Set up build.gradle for JavaFX and create Majima.txt if it is missing

Meowloid committed:
Change to run using the GUI but further fixes required
Basic functionality achieved

Meowloid committed:
Fix unintended behaviour of multiple erroneous commands stacking error messages

Meowloid committed:
Merge branch 'branch-Level-10'

Meowloid committed:
Fix Majimabot greeting

Meowloid committed:
Add speech bubbles to GUI

tohjh commented on src/Majima/Parser.java:
```diff
@@ -0,0 +1,96 @@
+        String args = (parts.length > 1) ? parts[1] : "";
+
+        switch (command) {
+            case "list":
```
there should be no indentation for switch clause

tohjh commented on src/Majima/Parser.java:
```diff
@@ -0,0 +1,96 @@
+    private static Command parseAddCommand(Task.TaskType type, String args) throws MajimaException {
+        String description;
+        switch (type) {
+            case TODO:
```
Shouldn't have indentation for all cases

Meowloid replied:
Fixed, thanks

tohjh commented on src/Majima/Majima.java:
```diff
@@ -0,0 +1,46 @@
+package Majima;
+
+import Majima.command.Command;
+import Majima.ui.Ui;
```
might want to consider sorting imports according to lexical order

Meowloid replied:
Fixed by right clicking the project directory in IntelliJ and optimizing imports from the dropdown menu/CTRL + ALT + O (windows)

tohjh commented on src/Majima/Parser.java:
```diff
@@ -0,0 +1,96 @@
+        String args = (parts.length > 1) ? parts[1] : "";
+
+        switch (command) {
+            case "list":
```
cases shouldn't have indentations

Meowloid replied:
See above

tohjh commented on src/Majima/Parser.java:
```diff
@@ -0,0 +1,96 @@
+            int index = Integer.parseInt(args);
+            return new DeleteCommand(index);
+        } catch (NumberFormatException e) {
+            throw new MajimaException("Eh? Kiryu-chan, that ain't no number!");
```
very cute messages from the chatbot!

Meowloid replied:
Thanks!

tohjh commented on src/Majima/task/TaskList.java:
```diff
@@ -0,0 +1,46 @@
+        this.tasks = new ArrayList<>(tasks);
+    }
+
+    public void addTask(Task task) {
```
consider putting javadocs for your public methods!

Meowloid replied:
Will do, in a later commit. Thanks!

tohjh commented on src/main/java/Majima/Parser.java:
```diff
@@ -0,0 +1,106 @@
+        }
+    }
+
+    private static int parseTaskNumber (String args) {
```
might want to remove the space between the function and parameter

Meowloid replied:
Fixed and pushed, thanks for catching this for me.

Meowloid committed:
Fix some style errors and order of imports

zzawook commented on src/Majima/Majima.java:
```diff
@@ -0,0 +1,46 @@
+            try {
+                String fullCommand = ui.readCommand();
+                //placeholder for ui.linegap
+                Command c = Parser.parse(fullCommand);
```
What about using more meaningful name for this variable? Like currentCommand, or even simply 'command'?

Meowloid replied:
Done

zzawook commented on src/Majima/Majima.java:
```diff
@@ -0,0 +1,46 @@
+import Majima.storage.Storage;
+import Majima.task.TaskList;
+
+public class Majima {
```
I like how the codes are well organized and divided into function. It really helps legibility.

Meowloid replied:
Thanks!

zzawook commented on src/Majima/Majima.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public static void main(String[] args) {
+        new Majima("data/Majima.txt").run();
```
Any reason why you used solid text instead of constant/final variable in other class?

Meowloid replied:
Whoops. Unrefactored code that got by me. Thanks for the heads up!

Meowloid committed:
Add title to javafx window

zzawook commented on src/Majima/command/Command.java:
```diff
@@ -0,0 +1,12 @@
+import Majima.ui.Ui;
+
+public abstract class Command {
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws MajimaException;
```
Perhaps you could add a javadoc of what this execute command does?

Meowloid replied:
Done

zzawook commented on src/Majima/command/Command.java:
```diff
@@ -0,0 +1,12 @@
+import Majima.task.TaskList;
+import Majima.ui.Ui;
+
+public abstract class Command {
```
Using command pattern to encapsulate different types of command is good

zzawook reviewed (commented):
Overall, the coding standard looks good. Just some nits and javadocs to be added for public methods, as other comment suggested as well. Keep it up!

Meowloid committed:
Tag Level-10 as well as add Javadocs to Command class

Meowloid committed:
Add assert checks for null tasks in Deadline, Event and Todo

Meowloid committed:
Add javadocs to a few methods

Meowloid committed:
Remove redundant files and follow suggestions made on GitHub

Meowloid committed:
Merge branch 'branch-A-CodeQuality'

Meowloid committed:
Revert "Merge branch 'branch-A-CodeQuality'"

This reverts commit f252d1fc2527066c707249972bfe1de30e0c765f, reversing
changes made to a4c21eff9c7ebe780188911fb030b136c6cb555f.

Meowloid committed:
Add C-Help functionality

Meowloid committed:
Refactor to not use magic strings

Meowloid committed:
Update personality of Majimabot

Meowloid committed:
Change comment in handleUserInput for clarity

Meowloid committed:
Merge branch 'branch-A-Assertions'

Meowloid committed:
Add FindTest class for testing FindCommand

The FindTest class is important as it cannot be tested
in a vacuum due to its nature, yet is an important
part of our implementation.

To this end, the FindTest class provides the smallest
abstraction of the Majimabot that does not involve the
GUI but is more concerned with the intended behaviour
of the FindCommand.

Meowloid committed:
Remove dead code from Ui class

Meowloid committed:
Update README.md for Week 6 task

Meowloid committed:
Fix compile error that I missed

Meowloid committed:
Add Ui.png and change README.md for week 6
